kong/init.lua

```diff
@@ -90,8 +91,56 @@ local set_current_peer = ngx_balancer.set_current_peer
 local set_timeouts     = ngx_balancer.set_timeouts
 local set_more_tries   = ngx_balancer.set_more_tries
 
+local plugins_map_version
+local plugins_map_semaphore
+
+local PLUGINS_MAP_CACHE_OPTS = { ttl = 0 }
+local PLUGINS_MAP_PAGE_SIZE = 1000
+
+local configured_plugins
 local loaded_plugins
 
+local function build_plugins_map(dao, version)
+  local map = {}
+
+  local rows, err, offset
+
+  -- postgres implements paging offsets as an incrementing integer
+  -- cassandra's driver uses the native protocol paging implementation
+  -- so we rely on the DAO to feed us back the offset with each iteration.
+  -- postgres requires an initial value of 1 to prevent a duplicate read of
+  -- the first page
+  if singletons.configuration.database == "postgres" then
+    offset = 1
+  end
+
+  -- iterate through a series of pages (PLUGINS_MAP_PAGE_SIZE at a time)
+  -- so as to avoid making a huge 'SELECT *' query on the proxy path during rebuilds
+  while true do
+    rows, err, offset = dao.plugins:find_page(nil, offset, PLUGINS_MAP_PAGE_SIZE)
+
+    if not rows then
+      return nil, tostring(err)
+    end
+
+    for _, row in ipairs(rows) do
+      map[row.name] = true
+    end
+
+    if offset == nil then
```

Comment:
Can this ever be `false`?

Reply:
yes

Reply:
`offset` can be `false`, `nil` or something else. ok.